### ISSUE TITLE
Clarification on "Overzealous input validation"

### ISSUE THREAD

**Issue by Echomo-Xinyu** [nus-cs2103-AY2223S1/forum#347]
Hi Prof! Our team is a bit unsure about what it means by "Overzealous input validation". (section screenshot as below) In the example of input `1234 5678 (HP) 1111-3333 (Office)` for a single phone field, what is the expected action we should be taking against input? How is warning different from a block in terms of the behaviour in this case?

<img width="722" alt="MicrosoftTeams-image" src="https://user-images.githubusercontent.com/25452497/196139316-473fc774-c657-4446-b06f-a00fc73e8f92.png">

**Comment by damithc** [comment]
@Echomo-Xinyu Good to see you are checking your tP product behavior against these criteria in this iteration itself (most do this too late). Let me answer part of the question first, and leave a bit of time for other classmates to weigh in if they have thoughts about this.

> How is warning different from a block in terms of the behaviour in this case?

Blocking is when you don't allow `1234 5678 (HP) 1111-3333 (Office)` to be saved if it doesn't match your validation criteria. 
Warning can be done in at least two ways:
1. Save the given number but display something like `This doesn't look like a typical phone number. If there is an error in the number, you can use the edit command to ....`. 
2. Before accepting, ask the user to confirm this entry is intentional and not a data entry error e.g., `This doesn't look like a typical phone number. Are you sure you want to save this as the phone number?`.

**Comment by deeyonn** [comment]
In this context, I would block if there are dependencies of the parameter `phone number` that requires it to be of a certain strict format. Else if the purpose of the parameter is solely for display purposes, I would warn.

**Comment by damithc** [comment]
> In this context, I would block if there are dependencies of the parameter `phone number` that requires it to be of a certain strict format. Else if the purpose of the parameter is solely for display purposes, I would warn.

Good point @deeyonn When deciding how strict we need to with the data format, we definitely need to factor in how the stored data is used later.

**Comment by seanflyyy** [comment]
Hi prof @damithc, if we are encouraged to warn rather than block, would we not need to take into account the different possibilities of inputs and handle them accordingly. Moreover, if we are accepting anything as input, would that not defeat the purpose of validation. It seems like we are being penalised for adding validation?

Would be great if you could clarify this! Thank you!

**Comment by damithc** [comment]
> Hi prof @damithc, if we are encouraged to warn rather than block, would we not need to take into account the different possibilities of inputs and handle them accordingly. Moreover, if we are accepting anything as input, would that not defeat the purpose of validation. It seems like we are being penalised for adding validation?

@seanflyyy validation (blocking user actions in particular)  has a trade-off with flexibility (and user's discretion).  
1. What validation rules would you use for the field telephone number? 
2. In this specific case, what's wrong with allowing user to store `1234 5678 (HP) 1111-3333 (Office)` under the telephone number? i.e., how would blocking it benefit the user?

**Comment by damithc** [comment]
@Echomo-Xinyu as hinted in the above discussion, allow flexibility in inputs when it can add value and doesn't get in the way of the correct operations of the software. For example, blocking inputs that can crash the app or mislead the user later is perfectly fine. At the same time, it's good if the application can prevent/minimize user errors. Often the case is that the prograg doesn't have enough information to decide if the 'unusual' input is a user error or intentional. That is where a friendly warning to the user can be useful so that the user can decide for him/herself. The ultimate test is which course of action will be in the best interest of the user i.e., maximizing the utility of the application while minimizing mishaps.
